compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs

```diff
@@ -365,14 +366,54 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
                 }
 
                 self.infcx.probe(|_| {
-                    if let Ok(_substs) = self.match_impl(impl_def_id, impl_trait_ref, obligation) {
+                    if let Ok(_) = self.match_impl(impl_def_id, impl_trait_ref, obligation) {
+                        if self.reject_fn_ptr_impls(impl_def_id, obligation) {
+                            return;
+                        }
                         candidates.vec.push(ImplCandidate(impl_def_id));
                     }
                 });
             },
         );
     }
 
+    fn reject_fn_ptr_impls(&self, impl_def_id: DefId, obligation: &TraitObligation<'tcx>) -> bool {
+        // Do not consider `impl<T: FnDef> Trait for T` for types other than function pointers
+        // or generic parameters that have a `FnDef` bound in scope.
+        for &(predicate, _) in self.tcx().predicates_of(impl_def_id).predicates {
+            let ty::PredicateKind::Clause(ty::Clause::Trait(pred))
+                = predicate.kind().skip_binder() else { continue };
+            if self.tcx().lang_items().fn_ptr_trait() != Some(pred.trait_ref.def_id) {
+                continue;
+            }
+            match pred.self_ty().kind() {
```

Comment:
my idea for the check here would be slightly different:

`subst` the `self_ty` using the `impl_substs`, if it is not an `FnPtr` or infer var, walk the `param_env` and reject eagerly here if no `whatever: FnPtr` predicate exists.